lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -494,36 +499,89 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
+            -- We only have stake pools since shelley era.
+            -- This is hardcoded for mainnet and will do nothing for other
+            -- networks.
+            -- TODO: query cardano-node for hard fork block (not yet possible)
+            --       https://jira.iohk.io/browse/CAD-1850
+-- | The very last block header of the Byron era, just before the transition
+--   to Shelley.
+-- 
+-- This is hard-coded for mainnet and will do nothing for other networks.
+--
+-- TODO: query cardano-node for hard fork block (not yet possible).
+-- https://jira.iohk.io/browse/CAD-1850
+--
+lastByronBlock :: BlockHeader
+lastByronBlock = BlockHeader
+    (SlotNo 4492799)
+    (Quantity 4490510)
+    (Hash $ unsafeFromHex
+        "f8084c61b6a238acec985b59310b6ecec49c0ab8352249afd7268da5cff2a457")
+    (Hash $ unsafeFromHex
+        "aa83acbf5904c0edfe4d79b3689d3d00fcfc553cf360fd2229b98d464c28e9de")
 
     getHeader :: CardanoBlock sc -> BlockHeader
     getHeader = toCardanoBlockHeader gp
 
     forward
         :: IORef EpochNo
+        -> IORef Int
         -> NonEmpty (CardanoBlock sc)
         -> (BlockHeader, ProtocolParameters)
         -> IO (FollowAction ())
-    forward latestGarbageCollectionEpochRef blocks (_nodeTip, _pparams) = do
-        atomically $ forM_ blocks $ \case
+    forward latestGarbageCollectionEpochRef chunkCounter blocks _ = do
+        atomically $ forML_ blocks forAllBlocks forLastBlock
+        pure Continue
+      where
+        {-# INLINE forAllBlocks #-}
+        forAllBlocks = \case
             BlockByron _ -> pure ()
             BlockShelley blk -> do
                 let (slot, certificates) = poolCertsFromShelleyBlock blk
                 let header = toShelleyBlockHeader getGenesisBlockHash blk
-                runExceptT (putPoolProduction header (getProducer blk))
-                    >>= \case
-                        Left e ->
-                            liftIO $ traceWith tr $ MsgErrProduction e
-                        Right () ->
-                            pure ()
+                handleErr (putPoolProduction header (getProducer blk))
                 unless (null certificates) $ do
                     -- Before adding new pool certificates to the database, we
                     -- first attempt to garbage collect pools that have already
                     -- retired.
                     garbageCollectPools slot latestGarbageCollectionEpochRef
                     putPoolCertificates slot certificates
-        pure Continue
+
+        {-# INLINE forLastBlock #-}
+        forLastBlock = \case
+            BlockByron blk -> do
+                c <- liftIO $ readIORef chunkCounter
+                when (c `mod` 50 == 0) $ do -- every 50k blocks, insert a header
+                    putByronHeader (toByronBlockHeader gp blk)
+                liftIO $ modifyIORef' chunkCounter (+1)
+            BlockShelley _ -> pure ()
+
+        {-# INLINE handleErr #-}
```

Comment:
Leave inlines to the compiler by default. Add a pragma if it appears to be necessary.